src/traits.rs

```diff
@@ -656,14 +656,44 @@ pub trait PrimInt
     + BitAnd<Output=Self>
     + BitOr<Output=Self>
     + BitXor<Output=Self>
-    + Shl<usize, Output=Self>
-    + Shr<usize, Output=Self>
+    + Shl<u32, Output=Self>
+    + Shr<u32, Output=Self>
     + CheckedAdd<Output=Self>
     + CheckedSub<Output=Self>
     + CheckedMul<Output=Self>
     + CheckedDiv<Output=Self>
+    + Sub<Output=Self>
     + Saturating
 {
+    type Signed : PrimInt;
+    type Unsigned : PrimInt;
+
+    /// Transmutes the integer into an unsigned integer of the
+    /// same size (loss less).
+    ///
+    /// # Examples
+    ///
+    /// ~~~
+    /// use num::traits::PrimInt;
+    ///
+    /// assert_eq!((8i32).bits_to_unsigned().bits_to_signed(), 8i32);
+    /// assert_eq!((-1i32).bits_to_unsigned().bits_to_signed(), -1i32);
+    /// ~~~
+    fn bits_to_unsigned(self) -> Self::Unsigned;
+
+    /// Transmutes an integer into a signed integer of the
+    /// same size (loss less).
+    ///
+    /// # Examples
+    ///
+    /// ~~~
+    /// use num::traits::PrimInt;
+    ///
+    /// assert_eq!((8i32).bits_to_unsigned().bits_to_signed(), 8i32);
+    /// assert_eq!((-1i32).bits_to_unsigned().bits_to_signed(), -1i32);
+    /// ~~~
+    fn bits_to_signed(self) -> Self::Signed;
+ 
```

Comment:
I've renamed them for now, will remove them later if the consensus says that we do not need them in `PrimInt` (I guess I will then propose another trait that implement these and requires `PrimInt`, which could be the way we end up doing this anyways even if consensus agrees we need them to avoid breaking the API). I've added a TODO to the original issue to track this.